### PR TITLE
GUVNOR-2557: Add keyboard navigation through the table

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/GUVNOR-2557

### CODE REVIEW

**Comment by manstis** [comment]
Related to https://github.com/droolsjbpm/drools-wb/pull/243

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+        final GridWidget selectedGridWidget = getSelectedGridWidget();
+        if ( selectedGridWidget == null ) {
+            return;
+        }
+
+        boolean handled = false;
+        boolean renderLayer = false;
+        final int nativeKeyCode = event.getNativeKeyCode();
+
+        if ( nativeKeyCode == KeyCodes.KEY_DELETE ) {
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java]: @manstis maybe it's a matter of style, but I really don't like this long if else if clauses. What about splits that in a selector + action?

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+            scrollSelectedCellIntoView();
+
+        } else if ( event.isDownArrow() ) {
+            handled = true;
+            renderLayer = adjustSelection( CellSelectionManager.SelectionExtensionDirection.DOWN,
+                                           event.isShiftKeyDown() );
+            scrollSelectedCellIntoView();
+        }
+
+        if ( handled ) {
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java]: maybe extract that for a method

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+            return false;
+        }
+        return selectedGridWidget.selectCell( rowCount - 1,
+                                              columnCount - 1,
+                                              false,
+                                              false );
+    }
+
+    boolean scrollSelectedCellIntoView() {
+        final GridWidget selectedGridWidget = getSelectedGridWidget();
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java]: maybe extract that in smaller methods

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        if ( dx != 0 || dy != 0 ) {
+            adjustViewportTransform( gridLayer.getViewport(),
+                                     new Point2D( dx,
+                                                  dy ) );
+        }
+        return true;
+    }
+
+    private double getCellScrollDeltaX( final GridWidget gridWidget ) {
+        final GridData gridModel = gridWidget.getModel();
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java]: smaller methods maybe?

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java (hunk 1 above)
Comment (manstis) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java]: @ederign I'm all ears.. what do you mean by "selector + action"?

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+            dx = bounds.getX() + bounds.getWidth() - gridCellX - column.getWidth();
+        } else if ( gridCellX <= bounds.getX() ) {
+            dx = bounds.getX() - gridCellX;
+        }
+
+        return dx;
+    }
+
+    private double getCellScrollDeltaY( final GridWidget gridWidget ) {
+        final GridData gridModel = gridWidget.getModel();
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java]: same as above

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java [hunk 6]
```diff
@@ -16,33 +16,458 @@
...
+        int maxUiRowIndex = findMaxUiRowIndex( origin );
+        int minUiColumnIndex = findMinUiColumnIndex( origin );
+        int maxUiColumnIndex = findMaxUiColumnIndex( origin );
+        int proposedUiRowIndex = findExtent( originUiRowIndex,
+                                             minUiRowIndex,
+                                             maxUiRowIndex );
+        int proposedUiColumnIndex = findExtent( originUiColumnIndex,
+                                                minUiColumnIndex,
+                                                maxUiColumnIndex );
+        switch ( direction ) {
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java]: same comment of long if, maybe place that code in enum.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java [hunk 7]
```diff
@@ -16,33 +16,458 @@
...
+        }
+
+        return edit( uiRowIndex,
+                     ci );
+    }
+
+    @Override
+    public boolean startEditingCell( final int uiRowIndex,
+                                     final int uiColumnIndex ) {
+        final GridData gridModel = gridWidget.getModel();
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java]: smaller methods?

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java [hunk 8]
```diff
@@ -16,33 +16,458 @@
...
-    protected boolean hasSelectionChanged( final List<GridData.SelectedCell> currentSelections,
-                                           final List<GridData.SelectedCell> originalSelections ) {
-        final List<GridData.SelectedCell> cloneCurrentSelections = new ArrayList<GridData.SelectedCell>( currentSelections );
-        final List<GridData.SelectedCell> cloneOriginalSelections = new ArrayList<GridData.SelectedCell>( originalSelections );
-        cloneCurrentSelections.removeAll( originalSelections );
-        cloneOriginalSelections.removeAll( currentSelections );
-        return cloneCurrentSelections.size() > 0 || cloneOriginalSelections.size() > 0;
+    private double getCellHeight( final int uiRowIndex,
+                                  final int uiColumnIndex ) {
+        final GridCell<?> cell = gridModel.getCell( uiRowIndex,
```
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java]: same comment as above

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java (hunk 1 above)
Comment (ederign) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java]: something like:

KeyDownHandler handler = EnumOrStaticFabricClass.findKeyDownHandler(event.getNativeCode)
if (handler != null){
      handler.handle();
     //if handled == true code.
}

In that way, we split if selection and handle code, making it easier to test/read.

what do you think?

Comment (manstis) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidgetKeyboardHandler.java]: OK, I understand.. making changes.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java (hunk 8 above)
Comment (manstis) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/selections/impl/BaseCellSelectionManager.java]: @ederign I split the cell height calculation into smaller methods; but Git has not detected the change and marked this comment as out of date...

**Comment by manstis** [comment]
@ederign Changes complete. Please take a second look.

**Comment by ederign** [comment]
thanks @manstis!!!! +1

**Pull request merged**